File: searching_algorithm/binary_search.cpp
//binary search first the list must be sorted


#include <bits/stdc++.h>

using namespace std;

int binary_search(int list[], int index){
    
    int length = sizeof(list) / sizeof(list[0]);
    cout << "length : " << length << endl;
    int first = 0;
    int last = length -1;
    cout << "last : " << last << endl;
    
    int mid_number = floor((first + last) / 2);
    cout << "Mid_number : " << mid_number << endl;
    
    while(index > 0){
        if(list[mid_number] == index){
        return mid_number;
    }
    if(list[mid_number] < index){
        first = mid_number + 1;
    }
    else{
        last = mid_number -1;
    }
    }
    
    return -1;
    
    
    
    
    
}


void verify(auto index){
    if(index != -1){
        cout << "there is index found at : " << index <<endl;
        
    }else{
        cout << "index was not found : " << endl;
    }
}


int main(){
    
    int linear_array[10] = {1,2,3,4,5,6,7,8,9,10};
    int  result = binary_search(linear_array, 7);
    
    verify(result);
    
    return 0;    
}

File: searching_algorithm/binary_search.dart


int? BinarySearch(List<int> list, int target){

  int first = 0;
  int last = list.length - 1;

  while(first < last){

    int middle = (first + last) ~/ 2; 

    if(target > list[middle]) first = list[middle] + 1;
    if(target < list[middle]) last = list[middle] - 1;

    if(list[middle] == target) return list[middle];


  }


  return null;

}



void main(){
  List<int> lists = [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30];
  print(BinarySearch(lists, 22));
}